pkg/kubelet/cm/internal_container_lifecycle_windows.go

```diff
@@ -20,30 +20,160 @@ limitations under the License.
 package cm
 
 import (
-	"k8s.io/api/core/v1"
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
 	utilfeature "k8s.io/apiserver/pkg/util/feature"
 	runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1"
 	"k8s.io/klog/v2"
 	kubefeatures "k8s.io/kubernetes/pkg/features"
 	"k8s.io/kubernetes/pkg/kubelet/winstats"
+	"k8s.io/utils/cpuset"
 )
 
 func (i *internalContainerLifecycleImpl) PreCreateContainer(pod *v1.Pod, container *v1.Container, containerConfig *runtimeapi.ContainerConfig) error {
-	if i.cpuManager != nil && utilfeature.DefaultFeatureGate.Enabled(kubefeatures.WindowsCPUAndMemoryAffinity) {
-		allocatedCPUs := i.cpuManager.GetCPUAffinity(string(pod.UID), container.Name)
-		if !allocatedCPUs.IsEmpty() {
-			var cpuGroupAffinities []*runtimeapi.WindowsCpuGroupAffinity
-			affinities := winstats.CpusToGroupAffinity(allocatedCPUs.List())
-			for _, affinity := range affinities {
-				klog.V(4).InfoS("Setting CPU affinity", "container", container.Name, "pod", pod.Name, "group", affinity.Group, "mask", affinity.MaskString(), "processorIds", affinity.Processors())
-				cpuGroupAffinities = append(cpuGroupAffinities, &runtimeapi.WindowsCpuGroupAffinity{
-					CpuGroup: uint32(affinity.Group),
-					CpuMask:  uint64(affinity.Mask),
-				})
+	if !utilfeature.DefaultFeatureGate.Enabled(kubefeatures.WindowsCPUAndMemoryAffinity) {
+		return nil
+	}
+
+	klog.Info("PreCreateContainer for Windows")
+
+	// retrieve CPU and NUMA affinity from CPU Manager and Memory Manager (if enabled)
+	var allocatedCPUs cpuset.CPUSet
+	if i.cpuManager != nil {
+		allocatedCPUs = i.cpuManager.GetCPUAffinity(string(pod.UID), container.Name)
+	}
+
+	var numaNodes sets.Set[int]
+	if i.memoryManager != nil {
+		numaNodes = i.memoryManager.GetMemoryNUMANodes(pod, container)
+	}
+
+	// set affinity based on available managers
+	var finalCPUSet map[int]struct{}
+
+	if i.cpuManager != nil && !allocatedCPUs.IsEmpty() && i.memoryManager != nil && numaNodes.Len() > 0 {
```

Comment:
Thanks for all the unit tests on the merging of sets.  There is quite a bit of logic here, that might be nice to have unit test to verify all this behavior (the three scenarios and with each manager on/off)

Reply:
agree, i'll try and add some

Reply:
I did some slight refactoring and added unit tests for this logic - PTAL

Reply:
Thanks!